Parser/myreadline.c

```diff
@@ -418,11 +423,13 @@ PyOS_Readline(FILE *sys_stdin, FILE *sys_stdout, const char *prompt)
     else {
         rv = (*PyOS_ReadlineFunctionPointer)(sys_stdin, sys_stdout, prompt);
     }
-    Py_END_ALLOW_THREADS
 
+    // gh-123321: Must set the variable and then release the lock before
+    // taking the GIL. Otherwise a deadlock or segfault may occur.
```

Comment:
How can a deadlock occur due to this variable not being `NULL`?

Reply:
Release lock then take GIL.
If you take GIL and then release lock there's a [deadlock](https://github.com/python/cpython/pull/123323#discussion_r1739368059).

Reply:
Ah the link you posted is exactly what I wanted to know. Thanks!